File: backend/app/schemas/automation.py
import uuid

from pydantic import BaseModel, ConfigDict


class AutomationActionBase(BaseModel):
    action_type: str
    action_config: dict = {}
    position: int


class AutomationActionCreate(AutomationActionBase):
    pass


class AutomationActionRead(AutomationActionBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    rule_id: uuid.UUID


class AutomationRuleBase(BaseModel):
    name: str
    trigger_type: str
    trigger_config: dict = {}
    is_enabled: bool = True


class AutomationRuleCreate(AutomationRuleBase):
    board_id: uuid.UUID
    actions: list[AutomationActionCreate] = []


class AutomationRuleRead(AutomationRuleBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    board_id: uuid.UUID
    created_by: uuid.UUID
    actions: list[AutomationActionRead] = []


class AutomationRuleUpdate(BaseModel):
    name: str | None = None
    trigger_type: str | None = None
    trigger_config: dict | None = None
    is_enabled: bool | None = None

File: frontend/src/types/automation.ts
export interface AutomationAction {
  id: string;
  rule_id: string;
  action_type: string;
  action_config: Record<string, unknown>;
  position: number;
}

export interface AutomationActionCreate {
  action_type: string;
  action_config: Record<string, unknown>;
  position: number;
}

export interface AutomationRule {
  id: string;
  board_id: string;
  name: string;
  trigger_type: string;
  trigger_config: Record<string, unknown>;
  is_enabled: boolean;
  created_by: string;
  actions: AutomationAction[];
}

export interface AutomationRuleCreate {
  name: string;
  trigger_type: string;
  trigger_config: Record<string, unknown>;
  is_enabled?: boolean;
  board_id: string;
  actions: AutomationActionCreate[];
}

export interface AutomationRuleUpdate {
  name?: string;
  trigger_type?: string;
  trigger_config?: Record<string, unknown>;
  is_enabled?: boolean;
}

export const TRIGGER_TYPES = [
  "card_moved_to_list",
  "card_created_in_list",
  "label_added",
  "due_date_approaching",
  "checklist_completed",
] as const;

export const ACTION_TYPES = [
  "add_label",
  "remove_label",
  "move_card_to_list",
  "assign_member",
  "mark_due_complete",
  "post_comment",
] as const;
